<a href="https://colab.research.google.com/github/basnendarekop/data_mining/blob/main/Data_Cancer_Program_KNN_dan_DT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
cd drive/My\ Drive/Colab Notebooks

/content/drive/My Drive/Colab Notebooks


In [4]:
ls

'Copy of dataset'/                              Klasifikasi_MNIST.ipynb
'Copy of kendaraan_cnn.ipynb'                   Klasifikasi_MNIST_new.ipynb
 data_cancer.csv                                Latihan/
 data_mining/                                   mnist_test.csv
 dataset_daun/                                  mnist_train.csv
'Dataset MNIST'/                                Multatuli.jpg
 iris_dataset_Contoh_Program_KNN_dan_DT.ipynb   Pertamina.jpg
 kapal.ipynb                                    RumahSakit.jpg
 KDA.jpg                                        Tes.jpg
 Klasifikasi_Cancer_FA.ipynb                    TLD.jpg
 Klasifikasi_Cancer_LDA.ipynb                   UAS/
 Klasifikasi_Cancer_PCA.ipynb                   Untitled0.ipynb
'klasifikasi daun.ipynb'


In [5]:
#import library yang dibutuhkan
import numpy as np
import pandas as pd

In [6]:
#baca dataset
dataset = pd.read_csv('data_cancer.csv')
dataset.head()

id diagnosis  ...  fractal_dimension_worst  Unnamed: 32
0    842302         M  ...                  0.11890          NaN
1    842517         M  ...                  0.08902          NaN
2  84300903         M  ...                  0.08758          NaN
3  84348301         M  ...                  0.17300          NaN
4  84358402         M  ...                  0.07678          NaN

[5 rows x 33 columns]

In [7]:
#cek ukuran dataset
print(dataset.shape)

(569, 33)


In [8]:
#cek apakah ada fitur yang kosong
dataset.isnull().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

In [9]:
dataset=dataset.drop(columns=['Unnamed: 32'])
print(dataset.head())
print(dataset.shape)

         id diagnosis  ...  symmetry_worst  fractal_dimension_worst
0    842302         M  ...          0.4601                  0.11890
1    842517         M  ...          0.2750                  0.08902
2  84300903         M  ...          0.3613                  0.08758
3  84348301         M  ...          0.6638                  0.17300
4  84358402         M  ...          0.2364                  0.07678

[5 rows x 32 columns]
(569, 32)


In [10]:
dataset.head()

id diagnosis  ...  symmetry_worst  fractal_dimension_worst
0    842302         M  ...          0.4601                  0.11890
1    842517         M  ...          0.2750                  0.08902
2  84300903         M  ...          0.3613                  0.08758
3  84348301         M  ...          0.6638                  0.17300
4  84358402         M  ...          0.2364                  0.07678

[5 rows x 32 columns]

In [106]:
#masukan dataset ke variabel X dan Y, X untuk fitur Y untuk labelnya
X=dataset.iloc[:,2:5].values
Y=dataset.iloc[:,1].values
print(X.shape)
print(Y.shape)

(569, 3)
(569,)


In [107]:
print(Y)

['M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M'
 'M' 'B' 'B' 'B' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M'
 'M' 'B' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'B' 'M' 'B' 'B' 'B' 'B' 'B' 'M'
 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B'
 'M' 'M' 'B' 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'M' 'M' 'B' 'M' 'M' 'M' 'B' 'B'
 'B' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'B'
 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'M' 'M' 'B' 'M' 'M' 'B' 'B' 'B'
 'M' 'M' 'B' 'M' 'B' 'M' 'M' 'B' 'M' 'M' 'B' 'B' 'M' 'B' 'B' 'M' 'B' 'B'
 'B' 'B' 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'B' 'B' 'B' 'M'
 'M' 'B' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'B'
 'M' 'M' 'M' 'B' 'M' 'B' 'M' 'B' 'B' 'B' 'M' 'B' 'B' 'M' 'M' 'B' 'M' 'M'
 'M' 'M' 'B' 'M' 'M' 'M' 'B' 'M' 'B' 'M' 'B' 'B' 'M' 'B' 'M' 'M' 'M' 'M'
 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'M' 'B' 'B' 'B' 'B' 'B' 'M' 'M' 'B' 'B' 'M'
 'B' 'B' 'M' 'M' 'B' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'B

In [108]:
print(Y[0:5])

['M' 'M' 'M' 'M' 'M']


In [109]:
#kita ubah data string ke bentuk integer untuk 2 kelas
from sklearn.preprocessing import LabelEncoder
labelencoder_Y = LabelEncoder()
Y = labelencoder_Y.fit_transform(Y)
print(Y[1:50])

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0
 1 1 1 1 1 1 1 1 0 1 0 0]


In [110]:
#kita bagi dataset menjadi data training dan data test
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.25, random_state = 5)
print(X_train.shape)
print(X_test.shape)


(426, 3)
(143, 3)


In [111]:
#kita lakukan fitur scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)
print(X_train[0,0:9])
print(X_test[0,0:9])

[ 1.92441392 -0.40539592  1.8412917 ]
[0.2701831  1.21509766 0.35049099]


In [112]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

#menggunakan logisticregression model
# classifier = LogisticRegression(random_state = 2)

#Using KNN
classifier_1 = KNeighborsClassifier(n_neighbors = 7)

#Using SVC method of svm class to use Support Vector Machine Algorithm
# classifier = SVC(kernel = 'linear', random_state = 0)

#Using GaussianNB
# classifier = GaussianNB()

#Using Decision Tree Algorithm
classifier_2 = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)

#Using Random Forest Classification algorithm
# classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)

In [113]:
#training data KNN
classifier_1.fit(X_train, Y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=7, p=2,
                     weights='uniform')

In [114]:
#training data DT
classifier_2.fit(X_train, Y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [115]:
#predict data test untuk mencari akurasi KNN
Y_pred_1 = classifier_1.predict(X_test)

In [116]:
#predict data test untuk mencari akurasi DT
Y_pred_2 = classifier_2.predict(X_test)

In [117]:
Y_pred_1 #prediksi KNN

array([1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0])

In [118]:
Y_pred_2 #prediksi DT

array([1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0])

In [119]:
Y_test

array([1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0])

In [120]:
from sklearn.metrics import accuracy_score
#akurasi KNN
accuracy_score(Y_test, Y_pred_1)

0.916083916083916

In [121]:
#akurasi DT
accuracy_score(Y_test, Y_pred_2)

0.9230769230769231

In [122]:
from sklearn.metrics import classification_report
#pred_report KNN
pred_report_1=classification_report(Y_pred_1, Y_test)
print(pred_report_1)

              precision    recall  f1-score   support

           0       0.95      0.91      0.93        92
           1       0.85      0.92      0.89        51

    accuracy                           0.92       143
   macro avg       0.90      0.92      0.91       143
weighted avg       0.92      0.92      0.92       143



In [123]:
#pred report DT
pred_report_2=classification_report(Y_pred_2, Y_test)
print(pred_report_2)

              precision    recall  f1-score   support

           0       0.95      0.92      0.94        91
           1       0.87      0.92      0.90        52

    accuracy                           0.92       143
   macro avg       0.91      0.92      0.92       143
weighted avg       0.92      0.92      0.92       143



In [124]:
#mencari confusion metric
from sklearn.metrics import confusion_matrix
cm_1 = confusion_matrix(Y_test, Y_pred_1) #KNN
cm_2 = confusion_matrix(Y_test, Y_pred_2) #DT

In [125]:
print(cm_1) #CM KNN

[[84  4]
 [ 8 47]]


In [126]:
print(cm_2) #CM DT

[[84  4]
 [ 7 48]]


**Analisis KNN dan Decision Tree**


*   Akurasi KNN lebih baik daripada Decision Tree jika jumlah fitur yang digunakan untuk klasifikasi banyak, misalnya 30. Ketika nilai K nya dinaikkan lebi. Ketika nilai K=1 maka akurasi KNN dan Decision tree sama yaitu 0.93, sedangkan ketika K=3 akurasi KNN naik menjadi 0.95. Ketika nilai K dinaikkan menjadi K=5,7 dst, akurasi KNN tetap 0.95 dan tidak naik. 
*  Ketika jumlah fitur yang digunakan untuk klasifikasi diturunkan menjadi 5 saja, akurasi DT lebih baik yaitu 0.92 sedangkan akurasi KNN hanya 0.91 walaupun nilai K=7.












